Fix deezer extractor

LucBerge opened the pull request:
### Description of the *pull request* and other information

This pull request fix the Deezer extractor with full song support and closes #1591.

- Add blowfish encipher
- Add deezer downloader
- Add full song support
- Add artist extractor
- Add track extractor
- Add episode extractor
- Add show extractor

### Before submitting this *pull request* I have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of the DeezerExtractor code and I am willing to release it under [Unlicense](http://unlicense.org/). [See youtube-dl](https://github.com/ytdl-org/youtube-dl/issues/31075).
- [x] I am not the original author of the Blowfish key computation and cipher code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence). [See youtube-dl](https://github.com/ytdl-org/youtube-dl/pull/13194).

### What is the purpose of my *pull request*?
- [x] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [x] Core bug fix/improvement
- [x] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

LucBerge committed:
Fix deezer extractor

- Add blowfish encipher
- Add deezer downloader
- Add full song support
- Add artist extractor
- Add track extractor
- Add episode extractor
- Add show extractor

LucBerge committed:
Code convention and flake8

LucBerge committed:
flake8

LucBerge commented:
What about the failed core tests ?

pukkandan commented:
Dont import legacy compat variables

LucBerge commented on yt_dlp/extractor/deezer.py:
```diff
@@ -1,92 +1,283 @@
+    traverse_obj,
 )
 
+sys.stdout.reconfigure(encoding='utf-8')
```
I'm having encoding issues without this line

LucBerge commented on yt_dlp/extractor/deezer.py:
```diff
@@ -1,92 +1,283 @@
+        for track in traverse_obj(response, ('results', 'data')):
+
+            if 'DIGITAL_RELEASE_DATE' in track:
+                release_date = track.get('DIGITAL_RELEASE_DATE').replace(' ', '')
```
Maybe `release_date = track.get('DIGITAL_RELEASE_DATE', '').replace(' ', '')`

LucBerge commented on yt_dlp/extractor/deezer.py:
```diff
@@ -1,92 +1,283 @@
+            if 'DIGITAL_RELEASE_DATE' in track:
+                release_date = track.get('DIGITAL_RELEASE_DATE').replace(' ', '')
+            else:
+                release_date = None
```
Maybe `release_date = track.get('DIGITAL_RELEASE_DATE', '').replace(' ', '')`

Grub4K commented:
> So I copy the function code in the file ?

No, you use the functions provided by Python, for example the `chr` builtin instead of `compat_chr`. Compat functions were workarounds which are no longer needed since support for old Python versions is not required.

The decryption process could potentially be done more cleanly if the calculation of the key happens inside the downloader and is not dependant on the `key` property being present.

Also Cryptodome provides an internal Blowfish implementation (`Cryptodome.Cipher.Blowfish`). It might be worth using that one and fallback only if Cryptodome not available, like doing already for AES inside `dependencies.py`

LucBerge committed:
Reviews

- Remove utf-8 encoding sys reconfiguration
- Do not use compat functions anymore
- Blowfish key calculation in the extractor
- Minor fixes

LucBerge commented:
> No, you use the functions provided by Python, for example the `chr` builtin instead of `compat_chr`. Compat functions were workarounds which are no longer needed since support for old Python versions is not required.

Done.

> The decryption process could potentially be done more cleanly if the calculation of the key happens inside the downloader and is not dependant on the `key` property being present.

If by "inside the downloader" you mean the `DeezerBaseInfoExtractor`, then it's done.

> Also Cryptodome provides an internal Blowfish implementation (`Cryptodome.Cipher.Blowfish`). It might be worth using that one and fallback only if Cryptodome not available, like doing already for AES inside `dependencies.py`

I'll do it later.

pukkandan commented on yt_dlp/extractor/deezer.py:
```diff
@@ -1,92 +1,269 @@
+
+    GW_LIGHT_URL = "https://www.deezer.com/ajax/gw-light.php"
+    GET_URL = "https://media.deezer.com/v1/get_url"
+    BLOWFISH_KEY = b"g4el58wc0zvf9na1"
```
How is this key obtained?

LucBerge replied:
From here : https://github.com/ytdl-org/youtube-dl/pull/13194/files#diff-301769c0cffc70e23f61421e11e68f1c77bf60db2f7547430bd3c4c894609699R40

ryan5453 replied:
Leaving the key in the repository may not be the best idea. Deezer has sent [DMCAs](https://github.com/github/dmca/blob/master/2021/02/2021-02-10-deezer.md) to many repos containing it in the past. It specifically mentions `"To do so, these repositories are sharing the private encryption keys retrieved maliciously to bypass Deezer's security measures to unlawfully download its music catalogue, directly from Deezer’s servers in total violation of our rights and those of our music licensors (phonographic producers, performing artists, songwriters and composers)."`

It might be a better idea to [dynamically generate the key](https://github.com/yne/dzr/issues/11) or just make users provide it themselves.

LucBerge replied:
@ryan5453 I agree, this is a good idea. Note that this key is not directly used to decrypt the song. It is used to compute a new key that will decrypt the song.

LucBerge replied:
@ryan5453 I have no idea what [this code](https://github.com/yne/dzr/issues/11#issue-1049202131) returns. I can't run it on my windows. Do you know how to generate this key in python directly ?

bashonly replied:
you could implement it in python with something like this:
```py
    def _get_blowfish_key(self, string):
        hex_str_list = re.findall(r'%5B(0x.{51})%5D', string)
        if len(hex_str_list) != 2:
            return
        list_a, list_b = [list(reversed(hex_str.split('%2C'))) for hex_str in hex_str_list]
        if len(list_a) != 8 or len(list_b) != 8:
            return
        final = []
        for i in range(8):
            final.append(list_a[i])
            final.append(list_b[i])
        return intlist_to_bytes([int(f, base=16) for f in final])

    explore_page = self._download_webpage('https://www.deezer.com/en/channels/explore', None)
    app_web_js = self._download_webpage(
        _search_regex(r'<script src="(https://[\w.-]+/cache/js/app-web\.[a-f0-9]+\.js)',
                      explore_page, 'javascript'), None, note='Downloading webapp JS')
    blowfish_key = self._get_blowfish_key(app_web_js)
```
...which adds an import:
```py
from ..utils import (
    intlist_to_bytes,
    sanitized_Request,
    sanitize_open,
)
```

gamer191 commented:
Related PRs: https://github.com/ytdl-org/youtube-dl/pull/13194 https://github.com/ytdl-org/youtube-dl/pull/22074

LucBerge commented:
> Unfortunately the linked repo was DMCAed when youtube-dl was DMCAed, but it might be available on the wayback machine (which seems to be down right now)

Yes it is : https://web.archive.org/web/20200917071235/https://github.com/LucBerge/youtube-dl but the `deezer_uncipher` branch is not.

LucBerge commented:
Regarding [this comment](https://github.com/ytdl-org/youtube-dl/pull/13194#issuecomment-383663591) (and a discussion I had with someone 😅) authentication is not used in my code. It explains why the `MD5_ORIGIN` key is not present and why `MP3_64`, `MP3_128` and `MP3_MISC` are the only format I have access to [even though all of them exists](https://github.com/ytdl-org/youtube-dl/pull/13194/files#diff-301769c0cffc70e23f61421e11e68f1c77bf60db2f7547430bd3c4c894609699R61). Authentication code must be added in the futur for better song quality.

@gamer191 GitHub don't want to unlock my repo, they proposed to remove it so I can fork it again. My commits are still on my local computer, I can push it to a new github repo but it can be DMCA again for the same reason youtube-dl was. So maybe it's not the best idea.

LucBerge commented on yt_dlp/extractor/deezer.py:
```diff
@@ -1,92 +1,269 @@
+        api_token, license_token = self.get_api_license_tokens(data_id)
+
+        #################################
+        # GET TRACK TOKENS AND PREVIEWS #
```
GET DIRECT STREAM

LucBerge committed:
Minor changes

- Remove useless impot
- Fix tracks number for extractors
- Fix comment

LucBerge committed:
Generate key dynamically

LucBerge committed:
Merge branch 'master' into deezer

Ryan5453 reviewed (approved):
I haven't actually tested this code and therefore can not confirm if it actually works or not, but the new key extractor (in theory) is definitely a much better implementation than hardcoding it for legal reasons.

LucBerge commented:
> I haven't actually tested this code and therefore can not confirm if it actually works or not, but the new key extractor (in theory) is definitely a much better implementation than hardcoding it for legal reasons.

You mean generating the key dynamically based on the code you provided?

Ryan5453 commented:
> > I haven't actually tested this code and therefore can not confirm if it actually works or not, but the new key extractor (in theory) is definitely a much better implementation than hardcoding it for legal reasons.
> 
> You mean generating the key dynamically based on the code you provided?

Yes. I'm saying you did a good job updating that.

LucBerge commented on yt_dlp/blowfish.py:
```diff
@@ -0,0 +1,309 @@
+from .compat import compat_struct_unpack, compat_struct_pack
```
I guess I should not import compat functions

pukkandan commented:
Due to reasons we have discussed above (especially https://github.com/yt-dlp/yt-dlp/pull/4935#discussion_r978580754), I don't think it is a good idea to merge this code. In theory, we would have some protection by not hard-coding the key, but even so, it is a very risky move for us.

Since our [plugin system](https://github.com/yt-dlp/yt-dlp#plugins) has now matured, I recommend distributing this yourself as a plugin instead. This way, any DMCA will only affect your plugin, and not the entire project. While there are no downloader plugins yet[^1], you can hack it in due to the flexibility of Python

```py
from yt_dlp.downloader import PROTOCOL_MAP
PROTOCOL_MAP['deezer'] = DeezerFD
```

[^1]: We are thinking of ways to make IEs able to define downloaders easier

LucBerge commented:
> Due to reasons we have discussed above (especially [#4935 (comment)](https://github.com/yt-dlp/yt-dlp/pull/4935#discussion_r978580754)), I don't think it is a good idea to merge this code. In theory, we would have some protection by not hard-coding the key, but even so, it is a very risky move for us.


> It might be a better idea to [dynamically generate the key](https://github.com/yne/dzr/issues/11) or just make users provide it themselves.

The key is now dynamically generated as suggested

> Since our [plugin system](https://github.com/yt-dlp/yt-dlp#plugins) has now matured, I recommend distributing this yourself as a plugin instead. This way, any DMCA will only affect your plugin, and not the entire project. While there are no downloader plugins yet, you can hack it in due to the flexibility of Python
> 
> ```python
> from yt_dlp.downloader import PROTOCOL_MAP
> PROTOCOL_MAP['deezer'] = DeezerFD
> ```

My goal in this PR is to share the extractor with others so people will be able to download too. Nobody will use my extractor as a plugin...

> PS: Also, any PR adding new downloader is currently being held on pause since we are discussing re-working how IEs declare FDs

Any link to the issue/discussion ?

pukkandan commented:
> Nobody will use my extractor as a plugin...

People who need it will find it. It can be added to the wiki and I could change even the current extractor warning to point to plugins if desired.


> > PS: Also, any PR adding new downloader is currently being held on pause since we are discussing re-working how IEs declare FDs
> 
> Any link to the issue/discussion ?

A lot of the discussion happens on discord b/w maintainers. https://github.com/yt-dlp/yt-dlp/pull/3048 is one possible implementation that came out of it.

coletdjnz commented:
Plugin repos can be tagged with [yt-dlp-plugins](https://github.com/topics/yt-dlp-plugins) to be discoverable. With the new plugin system we are aiming to make it a viable alternative to having extractor support in core. It's pretty new so not too many have heard of it yet nor is there many public plugins listed yet, so will take some time.

Also, after some investigation a while back, the origins of some of this code are potentially from DMCA'd sources and likely not under unlicense (it goes further back than the PR on youtube-dl), so we likely cannot include it anyway in it's current form.

LucBerge commented:
> So we likely cannot include it anyway in it's current form.

Ok. So in which form can you accept it ?
